packages/babel-template/src/literal.ts

```diff
@@ -45,41 +45,33 @@ function buildLiteralData<T>(
   tpl: Array<string>,
   opts: TemplateOpts,
 ) {
-  let names;
-  let nameSet: Set<string>;
-  let metadata;
-  let prefix = "";
+  let prefix = "BABEL_TPL$";
+
+  const raw = tpl.join("");
 
   do {
-    // If there are cases where the template already contains $0 or any other
-    // matching pattern, we keep adding "$" characters until a unique prefix
+    // If there are cases where the template already contains $$BABEL_TPL$0 or any other
+    // matching pattern, we keep adding "$$" characters until a unique prefix
     // is found.
-    prefix += "$";
-    const result = buildTemplateCode(tpl, prefix);
-
-    names = result.names;
-    nameSet = new Set(names);
-    metadata = parseAndBuildMetadata(formatter, formatter.code(result.code), {
-      parser: opts.parser,
-
-      // Explicitly include our generated names in the whitelist so users never
-      // have to think about whether their placeholder pattern will match.
-      placeholderWhitelist: new Set(
-        result.names.concat(
-          opts.placeholderWhitelist
-            ? Array.from(opts.placeholderWhitelist)
-            : [],
-        ),
+    prefix = "$$" + prefix;
+  } while (raw.includes(prefix));
```

Comment:
Instead of trying until we add enough `$`s, can we instead do this and only compute it once?

```js
let $count = 0;
const re = /\$+(?=\d)/g;
let m;
while (m = re.exec(tpl)) $count = Math.max($count, m[0].length);

const prefix = "$".repeat($count + 1);
```

Reply:
I guess `includes()` would be a bit faster? Of course it doesn't matter.

Reply:
The difference is that with includes we iterate over the string n times (where "n" is the final number of times we have to expand the prefix), while with the regexp loop only once (because the regexp alwags start after the previous match, and matches prefixes of any length).

Reply:
It seems that only in the last case (very long code and `$$$$$$0`) the regex is slightly faster, while in other cases `includes()` is faster.
It should be faster to use `BABEL_TPL$` as per the current code.
[perf](https://perf.link/#eyJpZCI6InRiZGlidGdvczZ3IiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IHRwbCA9IFtcbihgYWAucmVwZWF0KDEwMjQpK1wiJDBcIikucmVwZWF0KDE2KSxcbihgYWAucmVwZWF0KDEwMjQpK1wiJCQkJCQkMFwiKS5yZXBlYXQoMTYpLFxuKGBhYC5yZXBlYXQoMTAyNCkrXCIkMFwiKS5yZXBlYXQoNjQpLFxuKGBhYC5yZXBlYXQoMTAyNCkrXCIkJCQkJCQwXCIpLnJlcGVhdCg2NCksXG4oYGFgLnJlcGVhdCgxMDI0KjEyOCkrXCIkMFwiKS5yZXBlYXQoNjQpLFxuKGBhYC5yZXBlYXQoMTAyNCoxMjgpK1wiJCQkJCQkMFwiKS5yZXBlYXQoNjQpLFxuXVswXSIsInRlc3RzIjpbeyJuYW1lIjoiYSIsImNvZGUiOiIgIGxldCBwcmVmaXggPSBcIlwiO1xuXG4gIGRvIHtcbiAgICBwcmVmaXggPSBcIiRcIiArIHByZWZpeDtcbiAgfSB3aGlsZSAodHBsLmluY2x1ZGVzKHByZWZpeCkpOyIsInJ1bnMiOlsxOTAwMCwyMTAwMCwzMDAwLDExMDAwLDE4MDAwLDEzMDAwLDEwMDAsMjIwMDAsMTgwMDAsNTAwMCwyNjAwMCwxMTAwMCw2MDAwLDI0MDAwLDI0MDAwLDE0MDAwLDI3MDAwLDE2MDAwLDE2MDAwLDgwMDAsMjEwMDAsMjAwMDAsNzAwMCw0MDAwLDE1MDAwLDkwMDAsNzAwMCwyOTAwMCwxMDAwMCwyMTAwMCwzMDAwLDMyMDAwLDI0MDAwLDI2MDAwLDI0MDAwLDEwMDAsMjUwMDAsMjcwMDAsNTAwMCwyNDAwMCw5MDAwLDUwMDAsODAwMCw3MDAwLDE5MDAwLDE5MDAwLDI0MDAwLDE0MDAwLDI2MDAwLDYwMDAsNDAwMCwxODAwMCwzMDAwMCw5MDAwLDEyMDAwLDUwMDAsMTAwMCwxNjAwMCwxODAwMCwyNjAwMCw0MDAwLDE0MDAwLDE4MDAwLDEzMDAwLDEzMDAwLDE5MDAwLDcwMDAsMjQwMDAsMTMwMDAsMTMwMDAsMjAwMCwxNTAwMCw4MDAwLDUwMDAsMTEwMDAsMTUwMDAsMzAwMDAsMzAwMCwxMzAwMCw5MDAwLDE3MDAwLDgwMDAsMTkwMDAsNjAwMCwzMDAwLDcwMDAsMjIwMDAsMTIwMDAsNzAwMCwxMjAwMCwxODAwMCwyMzAwMCwxNzAwMCwyMDAwMCwxMTAwMCwzMDAwMCwyMDAwMCwxMjAwMCwyMjAwMCw4MDAwXSwib3BzIjoxNDU2MH0seyJuYW1lIjoiYiIsImNvZGUiOiJsZXQgJGNvdW50ID0gMDtcbmNvbnN0IHJlID0gL1xcJCsoPz1cXGQpL2c7XG5sZXQgbTtcbndoaWxlIChtID0gcmUuZXhlYyh0cGwpKSAkY291bnQgPSBNYXRoLm1heCgkY291bnQsIG1bMF0ubGVuZ3RoKTtcbmNvbnN0IHByZWZpeCA9IFwiJFwiLnJlcGVhdCgkY291bnQgKyAxKTsiLCJydW5zIjpbMTYwMDAsMTEwMDAsMjEwMDAsOTAwMCwxMzAwMCw4MDAwLDEwMDAsMTQwMDAsMTUwMDAsMjkwMDAsMTcwMDAsMTAwMCwxMDAwLDIxMDAwLDIxMDAwLDcwMDAsMjIwMDAsOTAwMCwxMDAwMCwyNDAwMCwxNzAwMCwxNDAwMCwxNzAwMCwxMDAwLDgwMDAsNjAwMCw2MDAwLDI2MDAwLDEwMDAsNjAwMCwyOTAwMCwxMDAwLDkwMDAsMjMwMDAsMjEwMDAsMjQwMDAsMjAwMDAsMjQwMDAsMjAwMCwxNDAwMCw3MDAwLDEwMDAsNTAwMCwxMDAwLDE1MDAwLDEyMDAwLDcwMDAsMTIwMDAsMTAwMCwyNDAwMCwxMDAwLDE2MDAwLDQwMDAsNjAwMCw5MDAwLDkwMDAsMTAwMCwxMzAwMCwxNDAwMCwyMzAwMCw4MDAwLDExMDAwLDMwMDAsNTAwMCw3MDAwLDE2MDAwLDMwMDAsMTYwMDAsMTEwMDAsNzAwMCwxMDAwLDIwMDAsNTAwMCwyMDAwLDI4MDAwLDEzMDAwLDI3MDAwLDI5MDAwLDEwMDAwLDYwMDAsMTMwMDAsMzAwMCwxMjAwMCwyMDAwLDEwMDAsMjIwMDAsMTIwMDAsOTAwMCw0MDAwLDYwMDAsMTYwMDAsMjEwMDAsMTMwMDAsMTgwMDAsMjkwMDAsMTAwMCwxODAwMCw4MDAwLDE3MDAwLDI1MDAwXSwib3BzIjoxMTgwMH1dLCJ1cGRhdGVkIjoiMjAyMy0wNC0wMVQxNDowNjozMi41ODhaIn0%3D)